### PR TITLE
Add doc for N-API Abstract Operations APIs

### PR DESCRIPTION
**Pull request by digitalinfinity**
First cut at API documentation for N-API operations that fall under the "Abstract Operations" section in the ECMA262 spec. 

Ref: #153

### CODE REVIEW

File: napi_abstract.md
Comment (jasongin) [inline napi_abstract.md]: JavaScript ~~Objects~~ values

File: napi_abstract.md
Comment (jasongin) [inline napi_abstract.md]: Coerce JavaScript _values_ to specific value types.
(I don't know what "fundamental" was supposed to mean here.)

File: napi_abstract.md
Comment (jasongin) [inline napi_abstract.md]: There is no default to `napi_object`. The default case now returns an error.

File: napi_abstract.md
Comment (jasongin) [inline napi_abstract.md]: ~~object~~ value
(many places in this doc)

**Review by jasongin — CHANGES_REQUESTED** [review]
Not all values are objects. :)

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Did a find-replace through the document 😄

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Reworded

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

**Comment by digitalinfinity** [comment]
Thanks @jasongin - pushed a commit that addresses your feedback

**Review by jasongin — APPROVED** [review]

File: napi_abstract.md
Comment (mhdawson) [inline napi_abstract.md]: Long line ?

File: napi_abstract.md
Comment (mhdawson) [inline napi_abstract.md]: long line (needs to be less than 80 chars)

File: napi_abstract.md
Comment (mhdawson) [inline napi_abstract.md]: long line

**Review by mhdawson — CHANGES_REQUESTED** [review]
Generally looks ok, but we need to adjust the line lengths to be <80 chars.

File: napi_abstract.md
Comment (mhdawson) [reply napi_abstract.md]: Note I only pointed out a few of the long lines, there are others.

**Comment by digitalinfinity** [comment]
@mhdawson thanks for reviewing- fixed the wrapping, PTAL- thanks!

**Comment by digitalinfinity** [comment]
Thanks @mhdawson - pushed a commit to address your feedback

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: hyphen between "passed" and "in"

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "... invoking *the* &#96;typeof&#96; *o*perator on ..."

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "invoking *the* &#96;instanceof&#96; *o*perator on ..."

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "... invoking *the* &#96;IsArray&#96; operation ..."

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "... is a~~n~~ N-API representation ... "

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: &#96;node::Buffer&#96;

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "... represents a~~n~~ TypedArray"

**Review by gabrielschulhof — CHANGES_REQUESTED** [review]
Just a few more nits.

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Isn't an N-API correct? I think its "a NAPI" but "an N-API" - IIRC the article depends on whether the first syllable makes a vowel sound, not whether the first letter is a vowel? But I'm not a native speak so I'll defer to @jasongin's command of the English language 😄

Comment (jasongin) [reply napi_abstract.md]: @digitalinfinity I agree with your usage, and so does http://english.stackexchange.com/questions/1016/do-you-use-a-or-an-before-acronyms

**Review by mhdawson — APPROVED** [review]
Looks good to me subject to Gabriel's nits.

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

**Comment by digitalinfinity** [comment]
@gabrielschulhof thanks for the review- pushed a new commit to address the last round of feedback

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "...~~apply~~perform..."

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "...~~Objects~~values."

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "~~Objects~~values"

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "~~N-API~~"

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "~~N-API~~"

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "~~N-API~~"

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "~~N-API~~"

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "... detecting *an* External ..."

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "~~N-API~~"

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: "~~N-API~~"

File: napi_abstract.md
Comment (gabrielschulhof) [reply napi_abstract.md]: Oh, right, it sounds like "enn-ay-pea-eye".

Comment (gabrielschulhof) [reply napi_abstract.md]: Oooh, but then it's like "... is enn enn-ay-pea-eye representation ..."

Perhaps we should say "Whether the given value holds a `node::Buffer`".

File: napi_abstract.md
Comment (gabrielschulhof) [inline napi_abstract.md]: backticks please

File: napi_abstract.md
Comment (gabrielschulhof) [reply napi_abstract.md]: Actually for the API below this one we have "Whether the given &#96;napi_value&#96; represents an Error object", so let's make this similar: "Whether the given &#96;napi_value&#96; represents a &#96;node::Buffer&#96; object".

**Review by gabrielschulhof — APPROVED** [review]
LGTM aside from a few minor issues.

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: Fixed

File: napi_abstract.md
Comment (digitalinfinity) [reply napi_abstract.md]: ok- changed

**Comment by digitalinfinity** [comment]
Thanks @gabrielschulhof - fixed all the nits - I'll leave the PR open for another day in case you want to take another look, but since you've already approved, I'll land this tomorrow if you have no additional feedback

**Comment by mhdawson** [comment]
landed as 2f8e42d102190cc46672bc297df684bf718ad97f

**Pull request closed without merging**
